pyoneer/rl/targets/reward_ops.py

```diff
@@ -0,0 +1,668 @@
+                   discounts=.99,
+                   lambdas=.975,
+                   weights=1.,
+                   time_major=False,
+                   back_prop=False,
+                   name=None):
+    """Computes lambda-returns.
+
+    For example, this can be used to compute TD(lambda):
+
+        ```
+        td_lambda = lambda_returns(r(t), V(t+1)) - V(t)
+        ```
+
+    Args:
+        rewards: tensor of shape [Batch x Time], [Batch x Time x ...].
+        values: tensor of shape [Batch x Time], [Batch x Time x ...].
+        bootstrap_value: tensor of shape [], [Batch], [Batch x ...].
+        discounts: tensor of shape [], [...], [Batch], [Batch, ...],
+            [Batch x Time], [Batch x Time x ...].
+        lambdas: tensor of shape [], [...], [Batch], [Batch, ...],
+            [Batch x Time], [Batch x Time x ...].
+        weights: tensor of shape [], [...], [Batch], [Batch, ...],
+            [Batch x Time], [Batch x Time x ...].
+        time_major: flag if tensors are time_major.
+            Batch and Time are transposed in this doc.
+        back_prop: allow back_prop through the calculation.
+        name: optional op name.
+
+    Returns:
+        tensor of shape [Batch x Time] or [Time x Batch]
+    """
+    with tf.name_scope(name or 'GeneralizedAdvantageEstimate'):
+        batch_axis = int(time_major)
+        time_axis = int(not time_major)
+
+        # Check rewards shape.
+        rewards = tf.convert_to_tensor(rewards)
+        rewards_rank = rewards.shape.rank
+        assert rewards_rank >= 2, 'rewards must be atleast rank 2.'
+        rewards_shape = tf.shape(rewards)
+
+        # Check discounts shape, broadcast to rewards shape.
+        discounts = tf.convert_to_tensor(discounts)
+        assert discounts.shape.rank <= rewards.shape.rank, (
+            'discounts rank must be <= rewards rank.')
+        discounts = tf.broadcast_to(discounts, rewards_shape)
+
+        # Check lambdas shape, broadcast to rewards shape.
+        lambdas = tf.convert_to_tensor(lambdas)
+        assert lambdas.shape.rank <= rewards.shape.rank, (
+            'lambdas rank must be <= rewards rank.')
+        lambdas = tf.broadcast_to(lambdas, rewards_shape)
+
+        # Check weights shape, broadcast to weights shape.
+        weights = tf.convert_to_tensor(weights)
+        assert weights.shape.rank <= rewards.shape.rank, (
+            'weights rank must be <= rewards rank.')
+        weights = tf.broadcast_to(weights, rewards_shape)
+
+        if bootstrap_value is None:
+            if time_major:
+                bootstrap_value = tf.zeros_like(values[-1, :])
+            else:
+                bootstrap_value = tf.zeros_like(values[:, -1])
+        else:
+            bootstrap_value = tf.convert_to_tensor(bootstrap_value)
+
+        bootstrap_value_t = tf.expand_dims(bootstrap_value, axis=time_axis)
+
+        # Shift values to tp1.
+        if time_major:
+            values_tp1 = values[1:, :]
```

Comment:
Could possibly use a more descriptive name than `values_tp1`; not sure what this mean (values from time plus one?)